### ISSUE TITLE
gradle init on Maven project results in StackOverflow

### ISSUE THREAD

**Issue by bonelli** [gradle/gradle#3078]
### Expected Behavior

`gradle init` invoked on the root project of my Maven multimodule should complete without errors or fail with meaningful output

### Current Behavior

`gradle init` invoked on the root project of my Maven multimodule will fail with a StackOverflow
```
bof@MBLMIXFBO /cygdrive/d/res-apps.trunk.svn/res-apps
$ gradle init --stacktrace
:wrapper UP-TO-DATE
:init
Maven to Gradle conversion is an incubating feature.
:init FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':init'.
> java.lang.StackOverflowError (no error message)

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':init'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.StackOverflowError
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:218)
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)
******** MANY MORE TIMES *******************
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)
        at org.gradle.buildinit.plugins.internal.maven.Maven2Gradle.generateFqn(Maven2Gradle.groovy:223)


* Get more help at https://help.gradle.org

BUILD FAILED in 2s
2 actionable tasks: 1 executed, 1 up-to-date
```

### Context

We have a modestly big multimodules Maven - Java/Spring project, with some (10ish) levels of pom projects. Maven builds are slow and this affects our productivity, I wanted to try gradle and I was just starting with `gradle init`.

### Steps to Reproduce (for bugs)

I cannot provide an example because our project is closed-sourced, and I have absolutely no hint about what caused the infinite recursion in those precise lines.
If someone can suggest me about what caused the infinite recursion I will forge a synthetic example.

### Your Environment

This is my environment versions, I'll be glad to provide more info.

```
$ gradle --version

------------------------------------------------------------
Gradle 4.2.1
------------------------------------------------------------

Build time:   2017-10-02 15:36:21 UTC
Revision:     a88ebd6be7840c2e59ae4782eb0f27fbe3405ddf

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_31 (Oracle Corporation 25.31-b07)
OS:           Windows 8.1 6.3 amd64
```

**Comment by bigdaz** [comment]
HI @bonelli. Thanks for reporting your issue.
The StackOverflow occurs in [this code](https://github.com/gradle/gradle/blob/b4e6b821865a87a1d63ed06d3842b9798bae55d8/subprojects/build-init/src/main/groovy/org/gradle/buildinit/plugins/internal/maven/Maven2Gradle.groovy#L214-L226), where `generateFqn()` is called recursively. Perhaps by inspecting the code and your project hierarchy you can work out what's triggering this issue.

Alternatively, you could add some logging to that method, build Gradle from source, and try running `gradle init`. Logging `project.parent.artifactId.text()` and `allProjects[0].artifactId.text()` might give a clue.

Unfortunately, I can't provide any more help without a reproducible example. I'll close this issue for now, but please reopen it or file another if you are able to glean some more information.

**Comment by bonelli** [comment]
Thanks @bigdaz I'll gladly look into it when I have a bit of spare time at work

**Comment by davidgraeff** [comment]
I have the same issue here. Gradle 4.3.1

**Comment by jschneider** [comment]
Same problem here with Gradle 4.4 (cf7821a6f79f8e2a598df21780e3ff7ce8db2b82)

**Comment by jschneider** [comment]
I created a test case that reproduces the problem

**Comment by bigdaz** [comment]
Thanks for the reproducible test case @jschneider, that should really help in tracking down the cause and finding a fix.

**Comment by blindpirate** [comment]
That would be fixed in `4.5`. Thank you all!
